### PR TITLE
WIP: Vala

### PR DESCRIPTION
**Pull request by albfan**
Work in progress for vala parser.

Written from scratch.

Right now is just a plain copy/paste from tcl parser

relates to #621

### CODE REVIEW

**Comment by albfan** [comment]
I think this is a wrong way to approach the implementation.

**Comment by albfan** [comment]
Reopening this as is the only acceptable approach to include vala into ctags.

File: parsers/vala.c [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+static void parseStatement (tokenInfo *const token)
+{
+	tokenInfo *lastToken = newValaToken ();
+	bool foundSignature = false;
+
+	do
+	{
+		tokenCopy (lastToken, token);
+		tokenRead (token);
+		if (tokenEqType (lastToken, 'class'))
```
Comment (k-takata) [inline parsers/vala.c]: Is `'class'` a mistake of `'c'`?

Comment (masatake) [reply parsers/vala.c]: You are correct.

Comment (albfan) [reply parsers/vala.c]: Check please, I think it is fixed right now. Not sure if that is the supposed right approach to keep going.

**Comment by albfan** [comment]
@masatake I have first question on this parsers:

How to detect a vala field (inside a class) and a property (field with `get` and `set`:

```
class Address {
   public string street;
...
}

class Person {
public Address address;
public string name {get; set;}
...
}
```

Understanding this seems the key to me. (maybe I should reread the links you suggest right now?)

**Comment by masatake** [comment]
> How to detect a vala field (inside a class) and a property (field with get and set:

Please show the pair of "input.vala" and "exepected.tags".  The pair is very important for me to understand what you need. If you cannot image "expected.tags" for the input, I cannot help you. In that case, what you have to do is reading ctags(1) man page carefully.

File: Units/parser-vala.r/class.vala.b/expected.tags [hunk 2]
```diff
@@ -0,0 +1,13 @@
+main	input.vala	/^void main(string[] args) {$/;"	method	line:8	language:Vala
+Address	input.vala	/^class Address {$/;"	class	line:15	language:Vala
+Person	input.vala	/^class Person {$/;"	class	line:23	language:Vala
+Address	input.vala	/^   public Address address {get; set;}$/;"	field	line:24	language:Vala
+_name	input.vala	16;"	field	line:16	language:Vala
+d_age	input.vala	17;"	field	line:17	language:Vala
+name	input.vala	16;"	property	line:16	language:Vala
```
Comment (albfan) [inline Units/parser-vala.r/class.vala.b/expected.tags]: @masatake this is a property.

class Person {
   public string name {get; set;}
}

at same time it is a field. In vala you can do something like

class Person {
   private int d_age;

   public int age {
      get { return d_age;}
      set { 
         if (value > 0) {
            d_age = value;
         } else {
            d_age = 0;
         }
      }
   }
}

`d_age` is a private field and `age` is the property.

Old implementation use `_<fieldname>` when there's no private field. Not sure if that really makes sense (pseudo tag or virtual tag?)

In general a field is:

<scope>* <type> <name>

if there's no scope it is private.

A property (can be more complex but is:

field {get; set;}

A property notifies its changes so you can add a listener, force to fill on constructor... things like that.

**Comment by albfan** [comment]
I added a review comment in `expected.ctags` that will be my approach to reimplement this, and once I see how to understand the scope for a tag (things inside a pair) things that depends on several previous tokens I will figure out myself  the rest.

Thanks for help here!

File: Units/parser-vala.r/class.vala.b/expected.tags (hunk 2 above)
Comment (masatake) [reply Units/parser-vala.r/class.vala.b/expected.tags]: No indentation breaks my small brain. Could you insert proper whitespaces?

**Comment by masatake** [comment]
I promised @k-tanaka I will focus on releasing related work till we can find a long term volunteer who drives the documentation related work of this project.

So I will show only sample code. I hope it  helps you.
Ideally, I should consider the work of your side.

**Comment by masatake** [comment]
See #621. I updated my branch.

**Comment by albfan** [comment]
@masatake I think I get it. I rebase this on top of your branch (I rebased on top of master too.

`parseClassBody` is really helpful. Let's see if I can complete the rest on my own.

**Comment by albfan** [comment]
Closing in favor of https://github.com/masatake/ctags/pull/6 where the development is really happening

**Pull request closed without merging**